dartagnan/src/main/java/com/dat3m/dartagnan/wmm/analysis/LazyRelationAnalysis.java

```diff
@@ -0,0 +1,651 @@
+            time(definition, start, System.currentTimeMillis());
+            return new RelationAnalysis.Knowledge(must, must);
+        }
+
+        @Override
+        public RelationAnalysis.Knowledge visitSetIdentity(SetIdentity definition) {
+            Filter filter = definition.getFilter();
+            long start = System.currentTimeMillis();
+            Map<Event, Set<Event>> data = program.getThreadEvents().stream()
+                    .filter(filter::apply)
+                    .collect(Collectors.toMap(e -> e, ImmutableSet::of));
+            EventGraph must = new ImmutableMapEventGraph(data);
+            time(definition, start, System.currentTimeMillis());
+            return new RelationAnalysis.Knowledge(must, must);
+        }
+
+        @Override
+        public RelationAnalysis.Knowledge visitInternal(Internal definition) {
+            long start = System.currentTimeMillis();
+            EventGraph must = new LazyEventGraph(visibleEvents, visibleEvents,
+                    (e1, e2) -> e1.getThread().equals(e2.getThread())
+                            && !exec.areMutuallyExclusive(e1, e2));
+            time(definition, start, System.currentTimeMillis());
+            return new RelationAnalysis.Knowledge(must, must);
+        }
+
+        @Override
+        public RelationAnalysis.Knowledge visitExternal(External definition) {
+            long start = System.currentTimeMillis();
+            EventGraph must = new LazyEventGraph(visibleEvents, visibleEvents,
+                    (e1, e2) -> !e1.getThread().equals(e2.getThread()));
+            time(definition, start, System.currentTimeMillis());
+            return new RelationAnalysis.Knowledge(must, must);
+        }
+
+        @Override
+        public RelationAnalysis.Knowledge visitProgramOrder(ProgramOrder definition) {
+            long start = System.currentTimeMillis();
+            Set<Event> domain = visibleEvents.stream().filter(e1 -> visibleEvents.stream()
+                            .anyMatch(e2 -> e1.getThread().equals(e2.getThread())
+                                    && e1.getGlobalId() < e2.getGlobalId()
+                                    && !exec.areMutuallyExclusive(e1, e2)))
+                    .collect(toSet());
+            Set<Event> range = visibleEvents.stream().filter(e2 -> visibleEvents.stream()
+                            .anyMatch(e1 -> e1.getThread().equals(e2.getThread())
+                                    && e1.getGlobalId() < e2.getGlobalId()
+                                    && !exec.areMutuallyExclusive(e1, e2)))
+                    .collect(toSet());
+            EventGraph must = new LazyEventGraph(domain, range,
+                    (e1, e2) -> e1.getThread().equals(e2.getThread())
+                            && e1.getGlobalId() < e2.getGlobalId()
+                            && !exec.areMutuallyExclusive(e1, e2));
+            time(definition, start, System.currentTimeMillis());
```

Comment:
In this particular example, there is probably not much gain. But in general, it is a good idea to have domain/range prefiltered like this, because you only store linear amount of information to avoid possibly quadratic amount of iteration. That's why the lazy graphs are defined using approximations of domain/range.